### PR TITLE
allow for garbage collection algo preference

### PR DESCRIPTION
**Pull request by ryphon**
https://theindiestone.com/forums/index.php?/topic/47417-dedicated-server-on-linux-vps-jre-crashes/

Only way to keep my server stable with 10+ people is to set to G1GC, but there's other java jre gc options

Afaik there's only a few java GC's that'd matter. ZGC is usually best, but G1GC can sometimes be more efficient with it's RAM usage. Probably related to something dumb in either my vm configs or otherwise, but seems like an easy change to allow it to be configurable more generally.

Maybe just adds tech debt to potential future work to deal with #38 and #23 but idk, this is an easy fix for my specific problems.

Signed-off-by: Dylan Shepard <dylan@shepard.dev>

### CODE REVIEW

**Comment by Renegade-Master** [comment]
Hey!

Thanks for the contribution!
If you are adding new variables, would you mind adding them to the [docker-compose.yaml](https://github.com/Renegade-Master/zomboid-dedicated-server/blob/0cd311bfb61f624b4dd042c17cf7c442d23a87d7/docker-compose.yaml) and [README.md](https://github.com/Renegade-Master/zomboid-dedicated-server/blob/0cd311bfb61f624b4dd042c17cf7c442d23a87d7/README.md), as well as the [run_server.sh](https://github.com/Renegade-Master/zomboid-dedicated-server/blob/0cd311bfb61f624b4dd042c17cf7c442d23a87d7/src/run_server.sh)?

That way it is documented, and available via all deployment methods.

**Comment by ryphon** [comment]
Updated with both fixed docs in run_server, as well as docker-compose and README updates @Renegade-Master

**Comment by Renegade-Master** [comment]
Thank you for adding the extra bits.
Looks like the developers have changed the successful shutdown message though.

Would you mind changing the line in [.github/workflows/docker-build.yml#L265](https://github.com/Renegade-Master/zomboid-dedicated-server/blob/b74e643901b664934bac447c8002445cbd4f6075/.github/workflows/docker-build.yml#L265) according to the following diff?

```diff
-          check_for_config "znet: Java_zombie_core_znet_SteamUtils_n_1Shutdown"
+          check_for_config "ZNet: SZombienet -> SSteamSDK: LogOff"
```

**Comment by ryphon** [comment]
Actions updated, sorry for the delay on this one!

**Comment by Renegade-Master** [comment]
Thanks for updating that! I would do it myself, but I have too many other things to be doing at the moment 😅

--- 

And it looks like the new versions of Project Zomboid have changed some configuration things.

image

That Port was definitely set, so something must have changed about the ini file structure.

image


I will try to take a look at this tomorrow. It should just be a case of figuring out what the new name for the main Port is.

**Comment by Renegade-Master** [comment]
@ryphon 
The required changes should be implemented in #43. When that is merged, we can rebase this PR, and hopefully get this merged in too.

File: README.md [hunk 1]
```diff
@@ -147,6 +147,7 @@ recommended for ease of configuration.
 | `GAME_VERSION`   | Game version to serve                        | [a-zA-Z0-9_]+     | `public`      |
 | `MAP_NAMES`      | Map Names (e.g. North;South)                 | map1;map2;map3    | Muldraugh, KY |
 | `MAX_RAM`        | Maximum amount of RAM to be used             | ([0-9]+)m         | 4096m         |
+| `GC_CONFIG`      | Specifices Java GC to use                    | ([0-9]+)m         | ZGC           |
```
Comment (Renegade-Master) [inline README.md:150]:
```suggestion
| `GC_CONFIG`      | Specifices Java GC to use                    | [a-zA-Z0-9_]+     | ZGC           |
```

**Comment by ryphon** [comment]
Hey I've been at Kubecon all week, anything you need from me here? Happy to rebase whatever.

**Comment by Renegade-Master** [comment]
I didn't realise there were two Kubecons per year. I went to the European one 6 months ago 😅

I just need to get this test working for the changed GC. It does the change, my syntax is just wrong for the validation. I couldn't get it last night.

**Review by Renegade-Master — APPROVED** [review]
Looks good to me anyway!

**Pull request merged**